### PR TITLE
Initial input

### PR DESCRIPTION
**Pull request by pastjean**
Things missing:
- [x] Tests
- [x] Decent metrics
- [x] No need to depend on github.com/pressly/chi for such a simple use case
- [x] dependency vendoring file using `go dep`
- [x] Travis-ci setup to build
- [x] Docker setup

https://github.com/coveo/k8s-aws-elb-tagger/blob/initial-input/README.md

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,3 @@
+MAINTAINER Pierre-Alexandre St-Jean <pa@stjean.me>
```
Comment (jocgir) [inline Dockerfile]: You really put a personal address?

File: web/middleware/request_logger.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	return logger
+}
+
+type wrappedResponse struct {
+	status int
+	bytes  int
+	http.ResponseWriter
+}
+
+var _ http.ResponseWriter = &wrappedResponse{}
```
Comment (jocgir) [inline web/middleware/request_logger.go]: What is the use of that code? Not sure to understand the side effect if there is.

File: main.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+							Key:   &k,
+							Value: &v,
+						})
+					}
+
+					addTagInput := &elb.AddTagsInput{
+						LoadBalancerNames: []*string{&elbName},
+						Tags:              awsTags,
+					}
+					if strings.ToLower(dry) == "true" {
```
Comment (lpbedard) [inline main.go]: I would put this once for all at dry definition:
`dry := strings.ToLower(os.Getenv("DRY_RUN")) == "true"` then use `if dry`

File: Makefile [hunk 4]
```diff
@@ -1,6 +1,9 @@
 VERSION ?= 1.0.0
 REVISION ?= git-$(shell git rev-parse --short HEAD)
 
+k8s-aws-elb-tagger: cmd/k8s-aws-elb-tagger/main.go
+	go build -o k8s-aws-elb-tagger cmd/k8s-aws-elb-tagger/main.go
```
Comment (jocgir) [inline Makefile]: I think that `go install ./cmd/k8s-aws-elb-tagger` would be preferable.

File: Makefile [hunk 5]
```diff
@@ -1,6 +1,9 @@
...
+
 k8s-aws-elb-tagger.linux.amd64: cmd/k8s-aws-elb-tagger/main.go
 	GOOS=linux GOARCH=amd64 go build -o k8s-aws-elb-tagger.linux.amd64 cmd/k8s-aws-elb-tagger/main.go
```
Comment (jocgir) [inline Makefile]: For cross environment, it is preferable to output result in a dedicated folder and add it in the .gitignore file.

File: Gopkg.lock [hunk 6]
```diff
@@ -0,0 +1,309 @@
+# This file is autogenerated, do not edit; changes may be undone by the next 'dep ensure'.
```
Comment (jocgir) [inline Gopkg.lock:1]: Do we really want to commit auto-generated files?

File: cmd/k8s-aws-elb-tagger/main.go [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+							Key:   &k,
+							Value: &v,
+						})
+					}
+
+					addTagInput := &elb.AddTagsInput{
+						LoadBalancerNames: []*string{&elbName},
+						Tags:              awsTags,
+					}
+					if strings.ToLower(dry) == "true" {
```
Comment (jocgir) [inline cmd/k8s-aws-elb-tagger/main.go]: Typically, environment variables like that are not requiring a specific value like "true". Any non null or zero value should be considered as true. The test should be made at the beginning of the function and dry should be a boolean variable.

File: cmd/k8s-aws-elb-tagger/main.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+						logger.Info("Tag To be added", "addTagInput", addTagInput)
+					} else {
+						elbAPI.AddTags(addTagInput)
+					}
+				}
+			}
+			time.Sleep(1 * time.Minute)
```
Comment (jocgir) [inline cmd/k8s-aws-elb-tagger/main.go]: I don't really encourage pooling when events are available.

File: web/middleware/request_logger.go (hunk 2 above)
Comment (pastjean) [reply web/middleware/request_logger.go]: i'll remove a lot from middleware this is a little trick that should be put in tests (which is my bad) 

it tests that `wrappedResponse` implements the interface `http.ResponseWriter`

File: main.go (hunk 3 above)
Comment (pastjean) [reply main.go]: i've reworked that part a little

File: Dockerfile (hunk 1 above)
Comment (pastjean) [reply Dockerfile]: oups automagic

File: Gopkg.lock (hunk 6 above)
Comment (pastjean) [reply Gopkg.lock:1]: This one is necessary by `go dep` to ensure we ahve the good versions when we install the dependencies

Comment (pastjean) [reply Gopkg.lock:1]: Here is more explanation: 
https://github.com/golang/dep/blob/master/FAQ.md#what-is-the-difference-between-gopkgtoml-the-manifest-and-gopkglock-the-lock

File: cmd/k8s-aws-elb-tagger/main.go (hunk 7 above)
Comment (pastjean) [reply cmd/k8s-aws-elb-tagger/main.go]: i should have made this a flag actually

File: k8s-aws-elb-tagger.go [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+		if err := r.refreshTags(); err != nil {
+			r.logger.Error("Failed to refresh tags", "err", err)
+		}
+
+		time.Sleep(r.waitBetweenRefresh)
+	}
+}
+
+func (r *tagRefresher) refreshTags() error {
+	services, err := r.k8sClient.CoreV1().Services("").List(v1.ListOptions{})
```
Comment (pastjean) [inline k8s-aws-elb-tagger.go:134]: i will refactor and make this part more clear and tested

File: k8s-aws-elb-tagger.go [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	// DefaultPort is the port the service is listening by default
+	DefaultPort = "3000"
```
Comment (vseguin) [inline k8s-aws-elb-tagger.go:27]: Just by curiosity, normal that we use PascalCase for these kinds of variables? Just wondering, it seems a bit out of place

File: k8s-aws-elb-tagger.go [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+
+	// ServiceAnnotationTagPrefix is the prefix to put on k8s elb service annotations
+	// to add tags to the AWS ELBs.
+	//
+	// ex:
+	//  k8s service annotation: aws-tag/owner="John Doe"
+	//  AWS tag will be: owner="John Doe"
+	ServiceAnnotationTagPrefix = "aws-tag/"
```
Comment (vseguin) [inline k8s-aws-elb-tagger.go:35]: Trailing slash really (and for others)? Should it be handled in your function instead? Seems error prone to me.

File: k8s-aws-elb-tagger.go (hunk 10 above)
Comment (jocgir) [reply k8s-aws-elb-tagger.go:27]: En Go, oui, c'est normal.

Comment (pastjean) [reply k8s-aws-elb-tagger.go:27]: Premiere lettre majuscule = variable exporté et utilisable à l'externe

Comment (jocgir) [reply k8s-aws-elb-tagger.go:27]: Dans terragrunt, l'auteur a utilisé le classique UPPERCASE_CONST pour les contantes, mais go sort une série de warnings.

File: k8s-aws-elb-tagger.go [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+
+	go refresher.Start()
+
+	port := GetenvOrDefault("PORT", DefaultPort)
+	logger.Info("Server starting", "port", port)
+	http.ListenAndServe(":"+port, Recoverer(logger)(m))
+	// FIXME: Implement graceful shutdown
+}
+
+func GetenvOrDefault(env, def string) string {
```
Comment (vseguin) [inline k8s-aws-elb-tagger.go]: Bizarre comme casing pour vrai

File: k8s-aws-elb-tagger.go (hunk 10 above)
Comment (vseguin) [reply k8s-aws-elb-tagger.go:27]: Okay, thx for clarifications!

File: k8s-aws-elb-tagger.go (hunk 11 above)
Comment (pastjean) [reply k8s-aws-elb-tagger.go:35]: like joining strings with a "/" ala java strings.join ?

File: k8s-aws-elb-tagger.go [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+	return tagsToApply
+}
+
+func kubernetesConfig() (*rest.Config, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		host, port := os.Getenv("KUBERNETES_HTTP_HOST"), os.Getenv("KUBERNETES_HTTP_PORT")
+
+		if len(host) == 0 || len(port) == 0 {
+			return nil, fmt.Errorf("Unable unable to load local proxy cluster configuration, KUBERNETES_HTTP_HOST & KUBERNETES_HTTP_PORT must be defined or if running in cluster KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT")
```
Comment (erocheleau) [inline k8s-aws-elb-tagger.go]: Unable unable 
:)

File: .travis.yml [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+- make test
+- make docker
+
+before_deploy: docker login -u="coveo+k8s_aws_elb_tagger" -p="$QUAY_TOKEN" quay.io
+
+deploy:
+- provider: releases
+  skip_cleanup: true
+  api_key:
+    secure: T7jqzA4hK+gg15l+OLJISPB/fduGVlbKCCHD0vvm9NdxAOpdZFjvI2Jm1vE4WITb/elkI3M20QMBYN253r80Up0AD5KYQvCkQac4YxM3yEb9jOxN18vL2enZEtN478bvwuXeY6q2hIt+memxAY9eBz2LitWIlCcr5peDxEzuMf1XVm95U1510HyeyMNU93aRMsBLXQ1J8p4oeRtB8MiYSP41a2sPmYjNy2jY0HNpTxrd+E6TbStu2ANQcGbK+7jliLaFI0K95vZRDoFi64aoL7uiwHeLMay3+I55XdcyxhtViPVUk+j0qBUKgTibXP8XzunJfD5H/Alq3YCQiBQd5KqiVO0zZhPNQsiIMYt9iKgECHMxCRiyDFC9hS4m1FqWuJwAUTVsCL+X6/Y7FinFhvje74hafv4isS+szcTJoyLWa81djYLceV00AIzC8roS4vdXPuN4gtVXJMlLAENRaksBRnvORn5TOviahKtxH3EqZpqPlBwcbGrOI/9F46DmBJbJEVxwLTBDj8JdyHOeMHhFED3MWzj9DwTSAYwWvD4aM+nCR/1KU7iGyJFhmmfBOD0eRUJtGqf4enZn6gvfpz390XjFlTUDqQr4XBB5q6T/4I+MnIImoI3+8Yfi+bJzum8aFUagGWhYMf86lx2WCuMPBNy4NOnvGMnoZKXGH4s=
```
Comment (vseguin) [inline .travis.yml:29]: Are those legit to commit?

File: README.md [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+```sh
+# Get dependency tool
+go get -u github.com/golang/dep/cmd/dep
+# Get the dependencies
+dep ensure
+```
+
+## Nice endpoints
+
+- /healthz
```
Comment (vseguin) [inline README.md:97]: healthz? :P

File: k8s-aws-elb-tagger.go (hunk 11 above)
Comment (vseguin) [reply k8s-aws-elb-tagger.go:35]: yes, something like that

File: k8s-aws-elb-tagger.go [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+	}
+
+	addTagInput := &elb.AddTagsInput{
+		LoadBalancerNames: []*string{&elbName},
+		Tags:              awsTags,
+	}
+	if r.dryRun {
+		r.logger.Info("Tag To be added", "addTagInput", addTagInput)
+	} else {
+		r.elbAPI.AddTags(addTagInput)
```
Comment (vseguin) [inline k8s-aws-elb-tagger.go]: Do we have any feedback if it fails when adding the tags?

Comment (pastjean) [reply k8s-aws-elb-tagger.go]: i should at least log error indeed

File: README.md (hunk 15 above)
Comment (pastjean) [reply README.md:97]: dunno its the default k8s endpoint used in softs :p

File: .travis.yml (hunk 14 above)
Comment (pastjean) [reply .travis.yml:29]: well its encrypted, scoped only to one docker repository. I guess its "ok"

**Pull request merged**